### PR TITLE
new format with file reference array

### PR DESCRIPTION
**Pull request by ticoann**
As we discussed in the meeting change the fwjr format to have file references. (to reduce the duplicate data)
(I haven't implemented the actual code yet)
1. fileArray will contains all the file nams (lfns and pfns) in a given document.
2. fileArrayRef will contain all the keys whose value is one or more files in fileArray.
3. values for the key which is in fileArrayRef will be replaces index of the fileArray to point the correct file name.

### CODE REVIEW

**Comment by ticoann** [comment]
@vkuznet, @yuyiguo, Valentin, Yuyi please review. Please let me know if there is a better way to do this (also better name for the new keys) Thanks.

**Comment by vkuznet** [comment]
Seangchan,
the change looks fine to me. But I found that given json files miss
meta data part

```
 "meta_data": {"agent_ver": "1.0.13.pre8",
               "fwjr_id": "100-0",
               "host": "vocms008.cern.ch",
               "ts": 1454617125},
```

Could you please add it.

Meanwhile, I inspected the json and found another way we can generalize it. So
far the fwjr_prod.json contains two chain steps, the cmsRun1 and stageOut1. We
also been told that we may have up-to five different chain steps. It means that
we'll need to put them into schema at some point.  But, if all chain steps are
alike, i.e. have the same structure (name, input, output and other
sub-structures) why we can't generalize chain steps as well. If this is the
case I propose we'll do it in the following way:

```
{"meta_data":{}, "data":[{chain1_dict}, {chain2_dict}]
```

Here the data part will contain list of chain dicts. Every chain dict
will have identical structure, e.g. name, input, output, etc.
For example, chain1_dict would be cmsRun1 one, the chain2_dict will be
stageOut1 and if we'll need to add other chain steps they will be naturally
added to data list.

Let me know your opinion.
Valentin.

**Comment by ticoann** [comment]
Hi Valentin, That would be fine with me. Actually that is the similar to what we had originally. except we need to remove the dictionary keys (cmsRun1, stageOut1, etc)
So the new structure would be (I changed the names)

```
{"meta_data":{}, 
 "fileArray": [....}
 "steps":[{"name": "cmsRun1",
               "fields": {same_structure}},
              {"name": "statgeOut1",
               "fields": {same_structure}}, ...]
"task":/
}
```

we are talking about the same structure right? :smile: 

Seangchan

**Comment by vkuznet** [comment]
Seangchan, yes, this is what I meant and for schema file you don't need to explicitly
list stateOut1, since cmsRun1 will represent data-structure for step entry.
Here is the schema layout:

```
{"meta_data":{}, 
 "fileArray": [....],
 "steps":[{"name": "cmsRun1", "fields": {same_structure}}],
 "task":"bla"
}
```

Please make changes that I can review and we can finally generate some data
with it.
Valentin.

**Comment by ticoann** [comment]
Hi Valentin, I made the change manually. I will change the code next week. Please review.

**Comment by yuyiguo** [comment]
@ticoann , @vkuznet 

All changes look good to me. I have two questions:
1. 'fileArrayRef': What is its propose here? Do we really need to define the keys?
2. 'fields' is this  layer necessary? 
Yuyi

**Comment by yuyiguo** [comment]
One more thing we need for the schema is Archive Collect steps. Seangchan mentioned that this step is different from others?  What does it look like?

**Comment by vkuznet** [comment]
Yuyi, the fileArrayRef is needed to understand at which chain given file was
used, i.e. was it input lfn or output lfn, etc.
Regarding fields. We have two possibility either store
{name:{}} or store {"name": name, "fields":{}} structures. Basically
former is implicit structure while latter is explicit one.
I don't have strong preference. But based on internal data-structure (within
cmsRun1) we always use former case, i.e. implicit structure and probably
should stick to consistent structures.

**Comment by ticoann** [comment]
Hi Yuyi, 
Valentin already mentioned the first question. About the fields, I don't really have preference. I just added the that due to the reason what Valentin said. But I think you have point that it might be not necessary structure.  So I can remove it, if we agree.

**Comment by ticoann** [comment]
Actually Yuyi is talking bout
{'name': cmsRun1, 'analysis': {}, 'input':{} ...} 
instead of 
{'name': cmsRun1, fields: {'analysis': {}, 'input':{} ...}}

I can remove the fields as Yuyi points out it would be another layer of hierarchy

**Comment by yuyiguo** [comment]
It looks all good to me.

**Comment by vkuznet** [comment]
Yes, I think it is redundant, since actual reference ids are within specific
chain step. So we|user will know what comes from what.

**Comment by vkuznet** [comment]
Agreed, remove fields.

**Comment by yuyiguo** [comment]
@ticoann When you work on the code, could you create a stand alone code or something like a unit test that can convert the current data in CouchDB into this new schema. Then we can use it to generate fwjr data for WMArchive using the existing data and you can take time to work on changing the Agent code.

**Comment by yuyiguo** [comment]
@amaltaro Can you take a look at the new schema too?

**Comment by ticoann** [comment]
@yuyi, yes I am working on the stand alone code. Just need to update the patch below. (But probably some time next week)
https://github.com/dmwm/WMCore/pull/6614

File: test/data/samples/fwjr_production.py [hunk 1]
```diff
@@ -116,16 +47,16 @@
...
+                         'inputPFNs': [1],
                          #TODO change to empty string from None
                          'location': '',
                          'merged': False,
                          'module_label': 'ALCARECOStreamDtCalib',
                          'ouput_module_class': 'PoolOutputModule',
                          'outputDataset': '/Cosmics/CMSSW_7_0_0_pre11-DtCalib-RECOCOSD_TaskChain_Data_pile_up_test-v1/ALCARECO',
-                         'outputLFN': '/store/unmerged/CMSSW_7_0_0_pre11/Cosmics/ALCARECO/DtCalib-RECOCOSD_TaskChain_Data_pile_up_test-v1/00000/ECCFE421-08CB-E511-9F4C-02163E017804.root',
-                         'outputPFN': 'root://eoscms.cern.ch//eos/cms/store/unmerged/CMSSW_7_0_0_pre11/Cosmics/ALCARECO/DtCalib-RECOCOSD_TaskChain_Data_pile_up_test-v1/00000/ECCFE421-08CB-E511-9F4C-02163E017804.root',
+                         'outputLFN': 2,
```
Comment (vkuznet) [inline test/data/samples/fwjr_production.py]: Seangchan, here you assign outputLFN to be int, while few lines above infputLFNs are type list(int). I want to double check if outputLFN will be singles? May be for safety, generalization and consistency we need to assign list(int) in both cases.

Comment (ticoann) [inline test/data/samples/fwjr_production.py]: We can generalize as list but I don't think there any case single step generates multiple outputs.
@amaltaro, Alan is there a case? Anyway, we can generalize if there if possibility later.

Comment (amaltaro) [inline test/data/samples/fwjr_production.py]: Yes, one step can output several outputs.
PS.: I'll give more feedback on this PR in the coming days

**Comment by ticoann** [comment]
Thank Alan, Valentin, I updated the schema with the list on outputLFN and outputPFN

**Comment by yuyiguo** [comment]
@ticoann How is the code to convert fwjr to WMArchive format? If anything I can help, please let me know.

**Comment by ticoann** [comment]
Hi Yuyi, 
Following is the function which convert the data
https://github.com/dmwm/WMCore/pull/6614/files#diff-1be2f4f865842be142f794cf9c1528f2R78

Sorry for the delay but I am trying to fix some bugs found in reqmgr2 for testbed. I can probably update later today or tomorrow morning.

**Pull request merged**